linux/devdoc/completion_port_linux_requirements.md

```diff
@@ -0,0 +1,180 @@
+{
+    int epoll_op;
+    SOCKET_HANDLE socket;
+    ON_COMPLETION_PORT_EVENT_COMPLETE event_callback;
+    void* event_callback_ctx;
+    EPOLL_BUFFER_DATA buffer_data;
+} COMPLETION_PORT_INGRESS_DATA;
+
+MOCKABLE_FUNCTION(, COMPLETION_PORT_HANDLE, completion_port_create);
+MOCKABLE_FUNCTION(, void, completion_port_dec_ref, COMPLETION_PORT_HANDLE, completion_port);
+MOCKABLE_FUNCTION(, void, completion_port_inc_ref, COMPLETION_PORT_HANDLE, completion_port);
+MOCKABLE_FUNCTION(, int, completion_port_post_queue, COMPLETION_PORT_HANDLE, completion_port, const COMPLETION_PORT_INGRESS_DATA*, ingress_data);
+MOCKABLE_FUNCTION(, void, completion_port_dequeue, COMPLETION_PORT_HANDLE, completion_port, SOCKET_HANDLE, socket);
+```
+
+### completion_port_create
+
+```C
+MOCKABLE_FUNCTION(, COMPLETION_PORT_HANDLE, completion_port_create);
+```
+
+`completion_port_create` creates and initializes the completion port module
+
+`completion_port_create` shall create a completion port object.
+
+`completion_port_create` shall create the epoll module by calling `epoll_create`.
+
+`completion_port_create` shall create a thread to handle the epoll events.
+
+On success `completion_port_create` shall return the allocated `COMPLETION_PORT_HANDLE`.
+
+If there are any errors then `completion_port_create` shall fail and return `NULL`.
+
+### completion_port_dec_ref
+
+```C
+MOCKABLE_FUNCTION(, void, completion_port_dec_ref, COMPLETION_PORT_HANDLE, completion_port);
+```
+
+`completion_port_dec_ref` handles the decrement of the reference count and freeing the memory if necessary
+
+`completion_port_dec_ref` shall decrement the reference count for `completion_port`.
+
+If the reference count reaches 0, `completion_port_dec_ref` shall do the following:
+
+- increment the flag signaling that the threads can complete.
+
+- close the epoll object.
+
+- close the thread by calling `ThreadAPI_Join`
+
+- then the memory associated with `completion_port` shall be freed.
+
+### completion_port_inc_ref
+
+```C
+MOCKABLE_FUNCTION(, void, completion_port_inc_ref, COMPLETION_PORT_HANDLE, completion_port);
+```
+
+`completion_port_inc_ref` handles the incrementing of the reference count of the module
+
+If `completion_port` is `NULL`, `completion_port_inc_ref` shall return.
+
+Otherwise `completion_port_inc_ref` shall increment the internally maintained reference count.
+
+### completion_port_post_queue
+
+```C
+MOCKABLE_FUNCTION(, int, completion_port_post_queue, COMPLETION_PORT_HANDLE, completion_port, const COMPLETION_PORT_INGRESS_DATA*, ingress_data);
+```
+
+`completion_port_post_queue` adds an ingress item to the post queue for epoll signaling
+
+If `completion_port` is `NULL`, `completion_port_post_queue` shall return a non-NULL value.
+
+If `ingress_data` is `NULL`, `completion_port_post_queue` shall return a non-NULL value.
+
+`completion_port_post_queue` shall allocate a thread data object
+
+`completion_port_post_queue` shall allocate a thread data object
```

Comment:
suppose allocating it twice (once here, once in line 112) is unintentional victim of copy&paste?